### PR TITLE
Add keyboard navigation to editors in shadow roots, by adding shadow dom support to focus code

### PR DESCRIPTION
**Pull request by tinydylan**
Related Ticket:

Description of Changes:
* Add keyboard navigation to editors in shadow roots, by adding shadow dom support to focus code
* Minor refactors

Pre-checks:
* [X] Changelog entry added
* [X] Tests have been added (if applicable)
* [X] Branch prefixed with `feature/` for new features (if applicable)
* [X] License headers added on new files (if applicable)

Review:
* [X] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/sugar/src/test/ts/browser/FocusTest.ts [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
 import * as Compare from 'ephox/sugar/api/dom/Compare';
 import * as Focus from 'ephox/sugar/api/dom/Focus';
 import * as Insert from 'ephox/sugar/api/dom/Insert';
 import * as Remove from 'ephox/sugar/api/dom/Remove';
 import * as Body from 'ephox/sugar/api/node/Body';
 import Element from 'ephox/sugar/api/node/Element';
 import * as Attr from 'ephox/sugar/api/properties/Attr';
+import { withShadowElement } from 'ephox/sugar/test/WithHelpers';
+import { Option, OptionInstances } from '@ephox/katamari';
+import tOption = OptionInstances.tOption;
```
Comment (lnewson) [inline modules/sugar/src/test/ts/browser/FocusTest.ts:11]: This line looks wrong?

Comment (tinydylan) [reply modules/sugar/src/test/ts/browser/FocusTest.ts:11]: I've used this a bunch of times. From the TS manual, it looks like it's supposed to be for `import x = require('y')`, but it works for this. 

It's probably a bad idea, so I'll remove it.

Comment (lnewson) [reply modules/sugar/src/test/ts/browser/FocusTest.ts:11]: Yeah, seems like a bug that even works. We also can't use `require` atm IIRC, so we definitely shouldn't be using that.

Comment (tinydylan) [reply modules/sugar/src/test/ts/browser/FocusTest.ts]: Fixed.

File: modules/sugar/src/test/ts/browser/FocusTest.ts [hunk 2]
```diff
@@ -31,3 +37,96 @@ UnitTest.test('FocusTest', () => {
...
+    Remove.remove(outerInput);
+  });
+});
+
+UnitTest.test('Focus.hasFocus in ShadowRoot', () => {
+  withShadowElement((shadowRoot, innerDiv, shadowHost) => {
+    const innerInput: Element<HTMLElement> = Element.fromTag('input');
+    Insert.append(innerDiv, innerInput);
+
+    const outerInput: Element<HTMLElement> = Element.fromTag('input');
```
Comment (ltrouton) [inline modules/sugar/src/test/ts/browser/FocusTest.ts:90]: Should `outerInput` be removed at the end of the function like the other unit tests above?

Comment (tinydylan) [reply modules/sugar/src/test/ts/browser/FocusTest.ts:90]: Yep. Nice catch.
Fixed.

**Pull request merged**